src/vs/workbench/services/positronConsole/browser/positronConsoleService.ts

```diff
@@ -1921,90 +1931,142 @@ class PositronConsoleInstance extends Disposable implements IPositronConsoleInst
 	 * Processes pending input.
 	 */
 	private async processPendingInput(): Promise<void> {
-		// If there isn't a pending input runtime item, return.
-		if (!this._runtimeItemPendingInput) {
+		// If we are already processing pending input and the `await` below allowed us to loop
+		// back into here, refuse to process anything else right now.
+		if (this._pendingInputState !== 'Idle') {
 			return;
 		}
 
-		// Get the index of the pending input runtime item.
-		const index = this.runtimeItems.indexOf(this._runtimeItemPendingInput);
+		this._pendingInputState = 'Processing';
 
-		// This index should always be > -1, but be defensive.
-		if (index > -1) {
-			this._runtimeItems.splice(index, 1);
+		try {
+			return await this.processPendingInputImpl();
+		} finally {
+			this._pendingInputState = 'Idle';
 		}
+	}
 
-		// Get the pending input lines and clear the pending input runtime item.
-		const pendingInputLines = this._runtimeItemPendingInput.code.split('\n');
-		this._runtimeItemPendingInput = undefined;
+	private async processPendingInputImpl(): Promise<void> {
+		// If there isn't a pending input runtime item, return.
+		if (!this._runtimeItemPendingInput) {
+			return;
+		}
 
 		// Find a complete code fragment to execute.
+		let code = undefined;
 		const codeLines: string[] = [];
+
+		// Get the pending input lines. We keep this up to date at every iteration so it always
+		// reflects the current state of the pending input, even after an `await`, which may have
+		// allowed the user to append more pending input code.
+		let pendingInputLines = this._runtimeItemPendingInput.code.split('\n');
+
 		for (let i = 0; i < pendingInputLines.length; i++) {
 			// Push the pending input line to the code lines.
 			codeLines.push(pendingInputLines[i]);
 
-			// Determine whether the code lines are a complete code fragment. If they are, execute
-			// the code fragment.
+			// Determine whether the code lines are a complete code fragment.
 			const codeFragment = codeLines.join('\n');
 			const codeFragmentStatus = await this.session.isCodeFragmentComplete(codeFragment);
+
+			// If we have been interrupted, then `clearPendingInput()` has reset
+			// `_runtimeItemPendingInput` and there is nothing for us to do.
+			if (this._pendingInputState === 'Interrupted') {
```

Comment:
@softwarenerd had a question about whether or not we need the `Interrupted` state here. His point was that maybe we could just check if `this._runtimeItemPendingInput == undefined` after each `await`, and if it was, treat that as the interrupt.

I thought about this a little more and determined we do still need it. In theory it is possible for this to happen during a single `await`:
- `clearPendingInput()` can be called, setting `this._runtimeItemPendingInput == undefined`
- `addPendingInput()` can be called, setting `this._runtimeItemPendingInput` to a new pending input

In that case, once we get control back we won't be able to tell that we got interrupted if we don't have this `'Interrupted'` state that gets fired from `clearPendingInput()`